# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/ArrayBufferViewObject.cpp
# Commit: 180daaed9cc8
# Full Hash: 180daaed9cc8c4691427f7b34361f9ce078944e3
# Author: Steve Fink <sfink@mozilla.com>
# Date: 2021-08-13 09:27:46
# Regressor Bug: 1720422
# File Overlap Count: 2
# Description:
#   Bug 1720422 - Fill in the rest of the ArrayBuffer+ArrayBufferView hierarchy, and base more ArrayBufferViewObject public APIs off of new API r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D119857
# ==============================================================================

diff -r 410e13fdf4e5 -r 180daaed9cc8 js/src/vm/ArrayBufferViewObject.cpp
--- a/js/src/vm/ArrayBufferViewObject.cpp	Thu Aug 12 20:44:15 2021 +0000
+++ b/js/src/vm/ArrayBufferViewObject.cpp	Thu Aug 12 20:44:16 2021 +0000
@@ -257,6 +257,21 @@
   return offset;
 }
 
+JS_PUBLIC_API uint8_t* JS::ArrayBufferView::getLengthAndData(
+    size_t* length, bool* isSharedMemory, const AutoRequireNoGC&) {
+  MOZ_ASSERT(obj->is<ArrayBufferViewObject>());
+  size_t byteLength = obj->is<DataViewObject>()
+                          ? obj->as<DataViewObject>().byteLength()
+                          : obj->as<TypedArrayObject>().byteLength();
+  *length = byteLength;  // *Not* the number of elements in the array, if
+                         // sizeof(elt) != 1.
+
+  ArrayBufferViewObject& view = obj->as<ArrayBufferViewObject>();
+  *isSharedMemory = view.isSharedMemory();
+  return static_cast<uint8_t*>(
+      view.dataPointerEither().unwrap(/*safe - caller sees isShared flag*/));
+}
+
 JS_PUBLIC_API JSObject* JS_GetObjectAsArrayBufferView(JSObject* obj,
                                                       size_t* length,
                                                       bool* isSharedMemory,
@@ -274,17 +289,9 @@
                                                        size_t* length,
                                                        bool* isSharedMemory,
                                                        uint8_t** data) {
-  MOZ_ASSERT(obj->is<ArrayBufferViewObject>());
-
-  size_t byteLength = obj->is<DataViewObject>()
-                          ? obj->as<DataViewObject>().byteLength()
-                          : obj->as<TypedArrayObject>().byteLength();
-  *length = byteLength;
-
-  ArrayBufferViewObject& view = obj->as<ArrayBufferViewObject>();
-  *isSharedMemory = view.isSharedMemory();
-  *data = static_cast<uint8_t*>(
-      view.dataPointerEither().unwrap(/*safe - caller sees isShared flag*/));
+  JS::AutoAssertNoGC nogc;
+  *data = JS::ArrayBufferView::fromObject(obj).getLengthAndData(
+      length, isSharedMemory, nogc);
 }
 
 JS_PUBLIC_API bool JS::IsArrayBufferViewShared(JSObject* obj) {